src/topology/order.lean

```diff
@@ -294,6 +294,21 @@ lemma is_open_coinduced {t : topological_space α} {s : set β} {f : α → β}
   @is_open β (topological_space.coinduced f t) s ↔ is_open (f ⁻¹' s) :=
 iff.rfl
 
```

Comment:
This lemma matches the style of `is_open_coinduced` above, so is probably a good idea:
```suggestion
lemma preimage_mem_nhds_coinduced [topological_space α] {π : α → β}
  {s : set β} {a : α} :
  s ∈ @nhds _ (topological_space.coinduced π ‹_›) (π a) → π ⁻¹' s ∈ 𝓝 a :=
begin
  letI := topological_space.coinduced π ‹_›,
  rw [mem_nhds_sets_iff, mem_nhds_sets_iff],
  rintros ⟨V, hVs, V_op, mem_V⟩,
  exact ⟨π ⁻¹' V, set.preimage_mono hVs, is_open_coinduced.mp V_op, mem_V⟩
end

```

Reply:
No, that's definitely a bad idea. It makes the lemma less general. Your version will work only when the topology on `β` is definitionally equal to the topology pushed by `π` from `α`.

Reply:
Sure, but if it's not definitionally equal they can extract a propositional goal with `convert preimage_mem_nhds_coinduced` or similar.

Note that you [can prove the "general"](https://github.com/leanprover-community/mathlib/pull/7554/files#r628885015) version with this lemma + `subst`. So I'd argue that saying this version is less general is somewhat like saying `add_comm : a + b = b + a` is less general than `add_comm (hc : c = a + b) (hd : d = b + a) : c = d`.

Reply:
I'm not convinced at all. With that reasoning you can "prove" anything using `convert iff.rfl` or whatever.

Reply:
Let's take a specific example from `data.nat.basic`. Do you suggest we replace
```lean
theorem eq_one_of_mul_eq_one_right {m n : ℕ} (H : m * n = 1) : m = 1 :=
eq_one_of_dvd_one ⟨n, H.symm⟩
```
by
```lean
theorem eq_one_of_mul_eq_one_right_eric {m : ℕ} (H : m = 1) : m = 1 := H

theorem eq_one_of_mul_eq_one_right {m n : ℕ} (H : m * n = 1) : m = 1 :=
begin
  convert eq_one_of_mul_eq_one_right_eric H,
  rw mul_comm at H,
  rw [eq_one_of_dvd_one ⟨m, H.symm⟩, mul_one]
end
```

Reply:
That's not an accurate characterization of my argument. My argument is that if you have a lemma of the form:
```lean
lemma foo {x y z} {p f} (h : x = f y z) : p x y z :=
```
then you should instead have stated it
```lean
lemma foo' {x y z} {p f} : p (f y z) y z :=
```
(or at least, proved this version first and provided the messier `foo` later only if you have a convincing use-case for it).

That doesn't apply to the snippet you posted, because `H : m * n = 1` is not in the form `y = f x` for some variable `y`.